transaction/transaction.go

```diff
@@ -5,3 +5,46 @@ package transaction
 type GDTxnFw struct {
 	// TODO: Add stuff as required
 }
+
+// Txn is a set of steps
+//
+// Nodes is a union of the all the TxnStep.Nodes
+type Txn struct {
+	Steps []Step
+	Nodes []string
+}
+
+// Do runs the transaction on the cluster
+func (t *Txn) Do() error {
+	//First verify all nodes are online
+	for range t.Nodes {
+		/*
+			if !Online(n) {
+				return error
+			}
+		*/
+	}
+
+	//Do the steps
+	for i, s := range t.Steps {
+		if s.Nodes[0] == All {
+			s.Nodes = t.Nodes
+		} else if s.Nodes[0] == Leader {
+			//s.Nodes[0] = LeaderName
+		}
+
+		if e := s.do(); e != nil {
+			t.undo(i)
```

Comment:
Post undo, I think you'd need to break here and should not continue for further steps